tests/DCPS/Compiler/optional/main.cpp

```diff
@@ -6,11 +7,87 @@
 
 #include <ace/ACE.h>
 #include <ace/Log_Msg.h>
+#include <gtest/gtest.h>
 
-#include <cstdlib>
+// Ensure opendds_idl is generating std::optional
+TEST(OptionalTests, Empty)
+{
+  optional::OptionalMembers empty;
+  EXPECT_FALSE(empty.opt().has_value());
+  empty.opt(12);
+  EXPECT_TRUE(empty.opt().has_value());
+}
+
+TEST(OptionalTests, SerializationSizeXCDR1)
+{
+  optional::OptionalMembers empty;
+  OpenDDS::DCPS::Encoding encoding;
+  encoding.kind(OpenDDS::DCPS::Encoding::KIND_XCDR1);
+  EXPECT_EQ(12, OpenDDS::DCPS::serialized_size(encoding, empty)); // TODO this is probably incorrect
+
+  optional::OptionalMembers notEmpty;
+  notEmpty.opt(12);
+  notEmpty.strOpt("Hello World");
+  notEmpty.seqOpt(std::vector{123, 456});
+  EXPECT_EQ(44, OpenDDS::DCPS::serialized_size(encoding, empty)); // TODO this is probably incorrect
+}
 
+TEST(OptionalTests, SerializationSizeXCDR2)
+{
+  optional::OptionalMembers empty{};
+  OpenDDS::DCPS::Encoding encoding;
+  encoding.kind(OpenDDS::DCPS::Encoding::KIND_XCDR2);
+  EXPECT_EQ(3, OpenDDS::DCPS::serialized_size(encoding, empty));
+
+  optional::OptionalMembers notEmpty;
+  notEmpty.opt(12);
+  notEmpty.strOpt("Hello World");
+  notEmpty.seqOpt(std::vector{123, 456});
+  EXPECT_EQ(44, OpenDDS::DCPS::serialized_size(encoding, empty)); // TODO this is probably incorrect
+}
+
+TEST(OptionalTests, SerializationXCDR1)
+{
+  OpenDDS::DCPS::Encoding encoding;
+  encoding.kind(OpenDDS::DCPS::Encoding::KIND_XCDR1);
+
+  OpenDDS::DCPS::Message_Block_Ptr b(new ACE_Message_Block(100000));
+  OpenDDS::DCPS::Serializer strm(b.get(), encoding);
+
+  optional::OptionalMembers msg;
+  msg.opt(12);
+  msg.strOpt("Hello World");
+  EXPECT_TRUE(strm << msg);
+
+ optional::OptionalMembers msg2;
+ EXPECT_TRUE(strm >> msg2);
+ EXPECT_TRUE(msg2.opt().has_value());
+ EXPECT_EQ(msg.opt().value(), msg2.opt().value());
+ EXPECT_FALSE(msg2.seqOpt().has_value());
+}
+
+TEST(OptionalTests, SerializationXCDR2)
+{
+  OpenDDS::DCPS::Encoding encoding;
+  encoding.kind(OpenDDS::DCPS::Encoding::KIND_XCDR2);
+
+  OpenDDS::DCPS::Message_Block_Ptr b(new ACE_Message_Block(100000));
+  OpenDDS::DCPS::Serializer strm(b.get(), encoding);
+
+  optional::OptionalMembers msg;
+  msg.opt(12);
+  msg.strOpt("Hello World");
+  EXPECT_TRUE(strm << msg);
+
+ optional::OptionalMembers msg2;
+ EXPECT_TRUE(strm >> msg2);
+ EXPECT_TRUE(msg2.opt().has_value());
+ EXPECT_EQ(msg.opt().value(), msg2.opt().value());
+ EXPECT_FALSE(msg2.seqOpt().has_value());
```

Comment:
A common pattern we typically use to test serialization correctness is explicitly comparing the resulting serialized data against an expected result. The expected result is hand-written based on the serialization rules described in the XTypes spec. This can prevent a misinterpretation of the rules in the spec being applied to both serialization and deserialization direction which can effectively cancel their effects out and still give the correct final C++ object's fields. Having an explicit expected result can help test serialization and deserialization separately.

You can refer to the xcdr test for examples: https://github.com/OpenDDS/OpenDDS/blob/master/tests/DCPS/Compiler/xcdr/xcdr.cpp#L464-L483